Improve debug and JSON output

samford opened the pull request:
The debug output includes a mess of information in the beginning that I have never found useful and it's a chore to constantly scroll past it.

More importantly, the debug output doesn't provide enough relevant information about what's happening in livecheck internally. This makes it more difficult than it should be for anyone who isn't intimately familiar with livecheck's behavior to understand what's actually happening in a given check.

Even for people who spend a lot of time working with livecheck, not having better debug information means that we have to waste time figuring out what livecheck is doing instead of it being immediately apparent from the debug output.

The main things that this PR does related to improving debug output are:

* Removes the unhelpful glut of information printed at the start of debug output (so we don't have to scroll past it anymore). [If anyone actually finds this information useful, I could comment out these lines instead of removing them, so you can selectively uncomment them as needed.]
* Prints more information about the current check (in addition to the existing list of matched strings and the list of versions)
  * The name of the formula
  * If livecheck is using a livecheckable for this check
  * The URL that's being tried, before any processing
  * The processed URL, if it differs from the original URL
  * The strategy being used for the URL
  * The regex being used from the livecheckable, if provided

With these changes, the output looks something like this:

```
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/a52dec.rb
Loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-livecheck/Livecheckables/a52dec.rb
Formula:         a52dec
Livecheckable?:  Yes

URL:             http://liba52.sourceforge.net/downloads.html
Strategy:        page_match
Regex:           /href=.*?a52dec-(\d+(?:\.\d+)+)\.t/

Matched Text on Page:
0.7.4, 0.7.3, 0.7.2, 0.7.0

Matched Versions:
0.7.4 => #<Version:0x00007fa0fd874c68 @version="0.7.4", @tokens=[#<Version::NumericToken 0>, #<Version::NumericToken 7>, #<Version::NumericToken 4>]>
0.7.3 => #<Version:0x00007fa0fd874c18 @version="0.7.3", @tokens=[#<Version::NumericToken 0>, #<Version::NumericToken 7>, #<Version::NumericToken 3>]>
0.7.2 => #<Version:0x00007fa0fd874b00 @version="0.7.2", @tokens=[#<Version::NumericToken 0>, #<Version::NumericToken 7>, #<Version::NumericToken 2>]>
0.7.0 => #<Version:0x00007fa0fd874a60 @version="0.7.0", @tokens=[#<Version::NumericToken 0>, #<Version::NumericToken 7>, #<Version::NumericToken 0>]>
a52dec : 0.7.4 ==> 0.7.4
```

[I'll be creating a PR to refactor the strategy setup (leading up to the strategy feature PR) and in the process I'll also be expanding the debug output to display the URL and regex used in the strategy, if they differ from what's passed into the strategy. This is useful information but it's easier to implement this after the strategies are refactored into classes.]

---

Past that, this also improves the JSON output a bit. The big issue was that errors and skipped/versioned formulae appeared as normal output before the JSON was printed. This integrates these into the JSON output, which seems to be more in keeping with the intention of the `--json` flag.

* Errors now appear in the JSON output like `{"formula":"acme","errors":["Unable to get versions"]}`
* Versioned formulae now appear in the JSON output like `{"formula":"ansible@2.8","versioned":true}`
* Skipped formulae now appear in the JSON output like `{"formula":"anttweakbar","skipped":true,"skip_msg":"Not maintained"}` (`skip_msg` is omitted if it's not provided)

This is an improvement but there are still some different types of errors that appear before the JSON output that we'll need to address in a different manner. For example:

```
fatal: repository 'https://git.postgresql.org/gitweb/?p=pg_top.git/' not found
warning: redirecting to https://git.b-ehlers.de/ehlers/unyaffs.git/
```

I'll work on this sometime in the future in a follow-up PR, unless someone has a good idea of how to address this. Basically, we want to capture the error and make it available in the JSON output like the other errors (e.g., `"errors":["fatal: repository 'https://git.postgresql.org/gitweb/?p=pg_top.git/' not found"]`).

The end goal is to not have any non-JSON output when the `--json` flag is in use (outside of maybe having a temporary progress indicator in the future that gets wiped out when the JSON is printed at the end).

---

I think this is an improvement overall but please check my work here, test this out for yourself, and leave any feedback about the debug and JSON output here.

nandahkrishna commented:
I tested this out locally for different situations (`livecheck` block, Livecheckable, no Livecheckable or block). It's quite relieving to look at the new output compared to the previous one.

The progress bar or something would be a great addition because the JSON takes a while when you're doing it for multiple formulae at once.

SMillerDev commented:
I don't have any comments on the code, but as someone who works with data imports a lot my personal preference would be to have error and skip messages consolidated in a `message` field. Combined with either the current booleans for error and skip or a `status` field I think that would make the json output easier to consume.

samford commented:
> The progress bar or something would be a great addition because the JSON takes a while when you're doing it for multiple formulae at once.

Yep, right now it's impossible to know what's going on when doing longer runs, so some people may think that the command isn't working. Before this PR, the printed errors at least let you know that progress was being made. This is something to tackle in a separate PR but I wanted to mention it, at least.


> my personal preference would be to have error and skip messages consolidated in a `message` field

That's a good point. My only concern is whether we could ever end up with more than one message (e.g., multiple errors or multiple lines of an error). How is this usually handled? Is `message` a string, an array of strings, or either of these things in different circumstances?

* With a string, we would seemingly have to combine messages with newlines, etc.
* With an array, the consumer would simply need to check the length and handle the message(s) appropriately. This allows the consumer to decide how to display more than one message, instead of being stuck with the formatting from the livecheck output (e.g., newlines) or having to split on a delimiter, which may be error prone.
* With a string or array setup, the consumer has to check the type and handle both situations (e.g., also checking the length of the array), so it would make everything more complicated and it's probably best to avoid this.

I'm in favor of using an array, even if there's only one message. With this setup, I imagine the name would be "messages" instead.

SMillerDev commented:
Yeah, array seems good to me.

samford commented:
Taking your other suggestion, I combined `error`, `skipped`, and `versioned` into one `status` field, which won't be present for a "good" result. With this and the `messages` addition, the output now looks like this:

* Error(s): `{"formula":"acme","status":"error","messages":["Unable to get versions"]}`
* Skipped: `{"formula":"anttweakbar","status":"skipped","messages":["Not maintained"]}`
* Versioned: `{"formula":"ansible@2.8","status":"versioned"}`

I think this will be a little more predictable and easier for consumers to work with, so I appreciate the feedback.

I also removed `is_` from the `is_outdated` and `is_newer_than_upstream` field names (to align them with the style of other boolean fields) and renamed `guessed` to `livecheckable` (to align it with the debug output).

---

I'm thinking about reducing the default JSON output to omit `livecheckable` (previously `guessed`) and `head`, since it's non-essential information. I would like to make more detailed information optionally available but it would be good to keep the default output minimal.

This would include the sort of information that we're now presenting in debug output like `livecheckable`, `url` (original, processed, strategy), `strategy`, and `regex` but not the matched text or versions, since that may be too much. This would allow us, for example, to write a script that takes the JSON output and tells us how many (and/or which) formulae use a given strategy.

It's probably appropriate to use the `--verbose` flag for this purpose, where we would include the additional JSON information if it's present.

---

Similarly, we may want to consider removing the ` (guessed)` text after the formula name in normal output and reserve this for the `--verbose` flag as well. Users who aren't familiar with livecheckables don't understand what "guessed" means, so it's either confusing or they don't care.

If there's an issue with a check where it's not working properly, we'll almost certainly run `--debug` anyway (to get a better idea of what's going on) and that will tell us whether there's a corresponding livecheckable.

nandahkrishna commented:
Correct me if I'm wrong here, @samford, but I think the `--verbose` option was used only to additionally print if a newer version was detected. This pretty much can be inferred from livecheck's visual output. In this PR you're printing `guessed` only if the flag is given.

Is it possible to remove this flag as it doesn't seem to be doing much else? And shift printing `guessed` to if `debug` is given?

Again I could be totally wrong here, but just a thought.

samford commented:
I'm happy to remove the pre-existing use of `--verbose` in `livecheck`. That is:

```ruby
if is_newer_than_upstream && Homebrew.args.verbose?
  opoo "#{formula_s} version is greater than the upstream version"
end
```

found at the end of `print_latest_version()` in `cmd/livecheck.rb`. This information is already presented in the normal output, where the formula version is displayed in red, so we don't get much value from that verbose output.

The `--verbose`, `--quiet`, and `--debug` flags have different purposes:

* `--verbose` isn't very useful before this PR but it becomes more useful afterward.
* `--quiet` prevents livecheck from displaying errors, much like `--newer-only` makes livecheck only display information for formulae with new versions.
* Currently, `--debug` is specifically used to show the debug output and does nothing else.

Using `--debug` in place of `--verbose` for the `(guessed)` output doesn't produce the same result. With the `--verbose` flag, you get the normal [terse] output with the addition of `(guessed)`. Using the `--debug` flag, you get the lengthy debug output. This includes the "Livecheckable?" information, so the `(guessed)` text would be useless in this context.

Regarding using `--debug` instead of `--verbose` for the additional JSON information, this requires internal changes to livecheck to work. Currently we use the debug flag to output the debug information, so we would have to change each of those checks from `if Homebrew.args.debug?` to `if Homebrew.args.debug? && !Homebrew.args.json?`. This is easy to do, so it comes down to whether we think `--debug` would make more sense to users than `--verbose`.

I think there's value in having a `--verbose` flag and we have one for other `brew` commands, so it's pretty standard. If we didn't use it at all, then I would maybe consider removing it but I think there's a use for it right now and potentially in the future, so it would be good to keep around.

If there's confusion about the purpose and difference between these flags, we may want to reword the help information a bit. For example, `--quiet` may be slightly better as "Avoid displaying errors or warnings". Otherwise, we can potentially explain livecheck's behavior in future documentation, so users properly understand how everything works.

nandahkrishna commented:
> remove the pre-existing use of `--verbose`

Yeah that seems good.

As for the confusion I had, it's more about how you can use (all) these flags together. Some of those combinations are pointless (as you'd said). Preventing useless combinations such as `--json` and `--debug`, `--verbose` and `--debug` or some others is too tedious, so we can just ignore that. Documentation would surely help in making things clear.

Thanks to your explanation, I'm convinced about the retention of the flags. I'm not sure what you had in mind but `--verbose` could possibly be used to print out some important information that we currently have as comments, which would be relevant to users of livecheck.

As for `--debug` versus `--verbose` for the extra information in the JSON, I view `--debug` as being analogous to `--json`. Both give you pretty similar, detailed information but in different formats. There are differences of course, but it seems slightly more logical to me that 'extra' details are provided when using `--verbose`.

samford commented:
These latest commits add the following:

* Remove the unnecessary `newer_than_upstream` verbose output in `livecheck.rb` (as the information is already conveyed in the normal output).
* Optionally display "Head only?" boolean information in the debug output (only if `formula.stable?` is false). This brings the debug output in line with the JSON output in this regard.
* Include some of the "debug" information in the verbose JSON output (e.g., original URL, processed URL, strategy, regex). This brings the verbose JSON output more in line with the debug output (the only difference being the omission of the lengthy "Matched text on page" and "Matched versions" information).

With these changes in place, this is what the default (non-verbose) JSON output looks:

```json
{
  "formula": "example",
  "version": {
    "current": "1.2.3",
    "latest": "1.2.3",
    "outdated": false,
    "newer_than_upstream": false
  }
}
```

This is what a full example of the verbose JSON output now looks like:

```json
{
  "formula": "example",
  "version": {
    "current": "1.2.3",
    "latest": "1.2.3",
    "outdated": false,
    "newer_than_upstream": false
  },
  "meta": {
    "livecheckable": true,
    "head_only": false,
    "url": {
      "original": "http://github.com/test/example/archive/v1.2.3.tar.gz",
      "processed": "http://github.com/test/example.git"
    },
    "strategy": "git",
    "regex": "/^v?(\\d+(?:\\.\\d+)+)$/i"
  }
}
```

[Note that `head_only` is only included here for demonstration purposes and would normally be omitted unless it's `true`.]

I'll be expanding the debug and verbose JSON output further in subsequent PRs but this is everything that I wanted to accomplish here. I would appreciate any final review before I merge this.

samford commented:
I figured out how to capture the Git errors but it's going to require changes to all strategy return values, so it's better to handle it in a follow-up PR.

I think this is good enough for now and everything else that I want to do requires changes that are beyond the scope of this PR. If anything about this can be improved further (the code, the debug output format, the JSON output format, etc.), we can handle it in subsequent PRs. If nothing else, we've at least laid the groundwork to get the debug information into `print_latest_version()`, where it can appear in JSON output.

Thanks to everyone for the feedback!
